core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala

```diff
@@ -110,9 +112,22 @@ class TaskMetrics private[spark] () extends Serializable {
    * joins. The value of this accumulator should be approximately the sum of the peak sizes
    * across all such data structures created in this task. For SQL jobs, this only tracks all
    * unsafe operators and ExternalSort.
+   * This is not equal to peakOnHeapExecutionMemory + peakOffHeapExecutionMemory
    */
+  // TODO: SPARK-48789: the naming is confusing since this does not really reflect the whole
+  //  execution memory. We'd better deprecate this once we have a replacement.
   def peakExecutionMemory: Long = _peakExecutionMemory.sum
 
+  /**
+   * Peak on heap execution memory as tracked by TaskMemoryManager.
+   */
+  def peakOnHeapExecutionMemory: Long = _peakOnHeapExecutionMemory.sum
+
+  /**
+   * Peak off heap execution memory as tracked by TaskMemoryManager.
+   */
+  def peakOffHeapExecutionMemory: Long = _peakOffHeapExecutionMemory.sum
```

Comment:
Discuss:
Is it required that `peakExecutionMemory` <= `peakOnHeapExecutionMemory + peakOffHeapExecutionMemory` ?
Any cases where this might get violated ?

I am trying to reason about completeness of these metrics (given we want to eventually deprecate the existing one).
I expect the above to hold, but want to make sure I am not missing anything.
+CC @JoshRosen

Reply:
peakExecutionMemory <= peakOnHeapExecutionMemory + peakOffHeapExecutionMemory?

I think yes, because`TaskMemoryManager.acquireExecutionMemory` is the only narrow waist for any execution memory acquisition and we maintain the memory here.

Instead, the legacy `peakExecutionMemory` is maintained in some operators (join, agg, sort), which is totally up to operator implementation.

Reply:
+1, I agree that the `peakExecutionMemory <= peakOnHeapExecutionMemory + peakOffHeapExecutionMemory` should hold:

If we trace through the existing callers of `incPeakExecutionMemory` it looks like all of the usages flow from counts that correspond to the acquireExecutionMemory waist.